docs: make installing a Downloaded Binary in the $PATH clearer (#6777)
ddev/ddev: docs/content/developers/building-contributing.md

Each [PR build](https://github.com/ddev/ddev/actions/workflows/pr-build.yml) creates GitHub artifacts you can use for testing, so you can download the one you need from the PR page, install it locally, and test using that build.

!!!tip "You can also [downgrade to an older version of DDEV](../users/usage/faq.md#how-can-i-install-a-specific-version-of-ddev) (perform a rollback)."

Normally, you can put any executable in your path, and it takes precedence, so you don't need to remove or disable an already installed DDEV instance, which we will use here. This example uses `~/bin`. Since not every OS has `$HOME/bin` in `$PATH`, you can create the folder and add it to your path by updating `~/.bashrc`, `~/.zshrc`, or another relevant shell configuration file with these commands:

```bash
mkdir ~/bin
echo 'export PATH="$HOME/bin:$PATH"' >>~/.bashrc
source ~/.bashrc
# Verify that `$HOME/bin` is the first entry in your `$PATH`
echo $PATH
```

> Choose a ZIP file for your OS and architecture, download it using `wget`, `curl`, `dl` or using a browser and unzip it, make it executable, and move it to the `~/bin` folder:

Author: Thinking more about this, the reason it irked me when I hovered over the links, was that I saw `[...]/2246879499.zip` which are "automagically" transformed.

I am not a fan of magic with computers, but in the end, clicking seems to be the easiest method.

It would work better for me, if there was a basic explanation of what it looks like, and what happens -- then I have no problem recommending the click-based solution. Something like this, recommending the click method, by moving it first and clarifying how it works:

```suggestion
Download a ZIP file for your OS and architecture by clicking the link (the result is something like `ddev-macos-arm64.zip`) or using `wget`, `curl`, `dl` and unzip it, make it executable, and move it to the `~/bin` folder:
```

```bash
# Example for macOS Apple Silicon:
unzip ddev-macos-arm64.zip
chmod +x ddev && mv ddev ~/bin/ddev
rm ddev-macos-arm64.zip
```

> ![Github Action PR Comment ZIP files](../images/github-action-pr-comment.png)

Author: Could a link to the nightly builds be added, in case someone wants to try out the "Testing a PR" process, and just needs a quick zip.file? I could easily see myself need this, at a later point in time, and not have to dig through PR's, to find a list of artifacts would be nice ...

```suggestion
![Github Action PR Comment ZIP files](../images/github-action-pr-comment.png)

Tip: If you need a zip-file to try out the "Testing a PR" process, see the [nightly builds](https://nightly.link/ddev/ddev/workflows/master-build/master).
```

Download and unzip the appropriate binary and place it in your `$PATH`.

!!!tip "You can also [downgrade to an older version of DDEV](../users/usage/faq.md#how-can-i-install-a-specific-version-of-ddev) (perform a rollback)."

> === "Homebrew with macOS or Linux"

Reviewer: At first i've thought this PR would have added the entire "homebrew with macos or linux" section but turns out after having the current and the new version side by side in preview that this section already existed. tbh  both sections ("homebrew  with macos or linux" and "installing a downloaded binary in the path" are sort of overlapping and doing basically the same. for me as a none technical user having more than one option where each is doing basically the same is rather stalling and freezing instead of providing clear and actionable instructions. i would rather suggest to drop the homebrew section entirely? plus the keyword homebrew is creating false expectations. my expectation based on the title would be the install of the PR binary would be based on homebrew but the sole reason why homebrew is named there is for unlinking ddev and linking it again after the testing is done.  plus by placing the binary in `/usr/local/bin/` sudo is needed. the other approach doesnt need it which would be preferable imho.

Reviewer 2: > sort of overlapping and doing basically the same.

I agree.

> plus by placing the binary in `/usr/local/bin/` sudo is needed. the other approach doesnt need it which would be preferable imho.

I want to mention that if you are using PhpStorm, the binary is initially saved for each project.

So if you have the original install at `/usr/local/bin/ddev` (or have a symlink here), add override it with `~/bin/ddev`, and open the project with PhpStorm, it will still use `/usr/local/bin/ddev` for its tasks, which is frustrating (because it will re-download `~/.ddev/bin/docker-compose` all the time).

---

Ideally I prefer to replace the original DDEV binary, and make a backup of the old one:

```
# Make yourself the binary owner and do a backup
binary="$(readlink -f "$(which ddev)")"
sudo chown "${USER}:${USER}" "${binary}"
cp "${binary}" "${binary}.bak"

# Download ZIP file using wget, curl, dl, or using a browser and unzip it
unzip ddev.zip
chmod +x ddev && mv ddev "${binary}"

# Do the testing, and to go back, do the following:
binary="$(readlink -f "$(which ddev)")"
cp "${binary}.bak" "${binary}"
```

I think this will be better in terms of interaction with PhpStorm.

I am already using this technique myself, happy with it, and I can call the original DDEV with `ddev.bak` at the same time.

Let's replace the Homebrew section with this technique.

    !!!warning "macOS and Unsigned Binaries"
        macOS doesn’t like these downloaded binaries, so you’ll need to bypass the automatic quarantine to use them:

        ```bash
        xattr -r -d com.apple.quarantine /usr/local/bin/ddev
        ```

        (The binaries on the master branch and the final release binaries _are_ signed.)

    After you’re done testing, you can delete your downloaded executable, and re-link the original Homebrew one:

    ```bash
    sudo rm /usr/local/bin/ddev
    brew link --force ddev
    ```

=== "Installing a Downloaded Binary in the `$PATH`"

    ### Installing a Downloaded Binary in the `$PATH`

    Normally, you can put any executable in your path, and it takes precedence, so you don't need to remove or disable an already installed DDEV instance, which we will use here. This example uses `~/bin`. `echo $PATH` and `which ddev` are valuable commands for debugging. Since not every distro has `$HOME/bin` in `$PATH`, you can create the folder and add it to your path in `~/.bashrc` with these commands:

    ```bash
    mkdir ~/bin
    echo 'export PATH="$HOME/bin:$PATH"' >>~/.bashrc

>     # Close and reopen your terminal, and verify that it worked with:

Reviewer: ```suggestion
    . ~/.zshrc
    # Verify that `~/bin` was added to your $PATH
```
instead of the instruction to close and reopen the terminal i would simply add a command to reload the .zshrc file inside the open terminal. and should a comment be added what `. ~/.zshrc` does as well?

Reviewer 2: Thank you, yes, sourcing the file should do the trick, in this case:

```
source ~/.bashrc
```

Edited your suggestion.

    echo $PATH
    ```

>     Next, download the ZIP archive, unzip it, make it executable, and move it to the `~/bin` folder:

Reviewer: strictly speaking i would expect the screenshot after line 96 right here in this context. here is the call to action to download the zip archive. i would also consider to remove line 90. on macos wget isnt available by default. it might be confusing if you have clear instructions for something that is not available for everyone. why not keep it simple and dont rely on a dl command here. people using commands like wget or curl use those anyway, other less technical users might get confused. if you are already providing the screenshot with the download links then also expect that someone simply clicks one of the links and downloads the necessary artifact?

Reviewer 2: Okay, we can replace wget line with a comment:

```
# Download ZIP file using wget, curl, dl, or using a browser and unzip it
```

    ```bash
    wget -O ddev.zip [LINK TO ZIP ARCHIVE]
    unzip ddev.zip

>     chmod +x ddev && mv ddev ~/bin/ddev

Reviewer: shouldnt the "macOS and Unsigned Binaries" box be added here as well (in particular in case the other sections would get dropped)?

    rm ddev.zip
    ```

>     Replace `[LINK TO ZIP ARCHIVE]` with the corresponding link from the PR comment:

Reviewer: according to my second to last comment i would drop this line here.

    ![Github Action PR Comment](../images/github-action-pr-comment.png)

>     Verify the replacement worked by running `ddev -v`. The output should be something like `ddev version v1.23.5-98-g3c93ae87e`, instead of the regular `ddev version v1.23.5`.

Reviewer: the clearer way for me validating everything is set up correctly is instead of using `ddev -v` ( i am using the version of ddev head anyway and it is not easy to distinguish what is a version of head and what is the version of the PR) i would simply run `which ddev` and make sure the path to ddev is located in `~/bin/ddev` (will be the absolute path instead of the ~) instead of `/opt/homebrew/bin/ddev` (in case you are on mac and using homebrew for example like in my case). and i just noticed the introductory paragraph of this section also references `which ddev` so i would definitely adjust this validation step and replace `ddev -v` with `which ddev`

    You need to run `ddev poweroff` and `ddev start` to download the Docker images that it needs.

>     After you’re done testing, you can delete your downloaded executable, restart your terminal, and again use the standard DDEV:

Reviewer: after you are done testing i would also recommend running ddev poweroff before deleting the artifact.

    Normally, you can put any executable in your path, and it takes precedence, so you don't need to remove or disable an already installed DDEV instance, which we will use here. This example uses `~/bin`. `echo $PATH` and `which ddev` are valuable commands for debugging. Since not every distro has `$HOME/bin` in `$PATH`, you can create the folder and add it to your path in `~/.bashrc` with these commands:

    ```bash
    mkdir ~/bin
    echo 'export PATH="$HOME/bin:$PATH"' >>~/.bashrc
    source ~/.bashrc
    # Verify that `~/bin` was added to your $PATH
    echo $PATH
    ```

    Next, download the ZIP archive, unzip it, make it executable, and move it to the `~/bin` folder:

    ```bash
    wget -O ddev.zip [LINK TO ZIP ARCHIVE]
    unzip ddev.zip
    chmod +x ddev && mv ddev ~/bin/ddev
    rm ddev.zip
    ```

    Replace `[LINK TO ZIP ARCHIVE]` with the corresponding link from the PR comment:

    ![Github Action PR Comment](../images/github-action-pr-comment.png)

    Verify the replacement worked by running `ddev -v`. The output should be something like `ddev version v1.23.5-98-g3c93ae87e`, instead of the regular `ddev version v1.23.5`.

>     You need to run `ddev poweroff` and `ddev start` to download the Docker images that it needs.

Reviewer 3: ```suggestion
    When DDEV detects a version change, it recommends powering down all running containters. Then, it will download the new images, if required.
```

DDEV does not need to "poweroff" to get new images:

```shell 
$ ddev start
You seem to have a new DDEV version.
During an upgrade it's important to `ddev poweroff`.
May I do `ddev poweroff` before continuing?
This does no harm and loses no data. [Y/n] (yes): n
Starting cypress...
20241122_stasadev_php_pecl: Pulling from ddev/ddev-webserver
132e290b53a3: Download complete
Digest: sha256:1111b1a5766693de8b432149941433a432c8d53f41c24ba1c1efd443a6e6f912
Status: Downloaded newer image for ddev/ddev-webserver:20241122_stasadev_php_pecl
```

### Installing a Downloaded Binary in the `$PATH`

Normally, you can put any executable in your path, and it takes precedence, so you don't need to remove or disable an already installed DDEV instance, which we will use here. This example uses `~/bin`. `echo $PATH` and `which ddev` are valuable commands for debugging. Since not every distro has `~/bin` in `$PATH`, you can create the folder and add it to your path.

Add `~/bin` to your path, by adding this in your `~/.bashrc` file:
```

> export PATH="~/bin:$PATH"

Reviewer 2: This should be `$HOME/bin`, not `~/bin`, because:

- #6722

Let me fix it.

Author: Great, feel free to update it.

```
Create the `~/bin` folder:
```
mkdir ~/bin
```
Close and reopen your terminal, and verify that it worked with `echo $PATH`.

> Next, download the executable file inside the `~/bin` folder, unzip it, and make it executable:

Reviewer 2: We download a ZIP archive, not a file.
It's better to add a real example with `wget` and `unzip`, because nobody probably uses that. (I will add it.)

Author: Sure, I am open for corrections and suggestions.
